pkg/server/server.go

```diff
@@ -327,8 +402,33 @@ func NewServerWithOpts(opts ...OpenFGAServiceV1Option) (*Server, error) {
 		graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
 	)
 
-	cycleDetectionCheckResolver.SetDelegate(localChecker)
-	localChecker.SetDelegate(cycleDetectionCheckResolver)
+	if s.rateLimitedCheckResolverEnabled {
+		rateLimitConfig := graph.RateLimitedCheckResolverConfig{
+			TimerTickerFrequency: s.rateLimitedCheckResolverTimerTickerFrequency,
+			MediumPriorityLevel:  s.rateLimitedCheckResolverMediumPriorityLevel,
+			MediumPriorityShaper: s.rateLimitedCheckResolverMediumPriorityShaper,
+			LowPriorityLevel:     s.rateLimitedCheckResolverLowPriorityLevel,
+			LowPriorityShaper:    s.rateLimitedCheckResolverLowPriorityShaper,
+		}
+
+		s.logger.Info("Enabling dispatch throttling",
+			zap.Duration("TimerTickerFrequency", s.rateLimitedCheckResolverTimerTickerFrequency),
+			zap.Uint32("MediumPriorityLevel", s.rateLimitedCheckResolverMediumPriorityLevel),
+			zap.Uint32("MediumPriorityShaper", s.rateLimitedCheckResolverMediumPriorityShaper),
+			zap.Uint32("LowPriorityLevel", s.rateLimitedCheckResolverLowPriorityLevel),
+			zap.Uint32("LowPriorityShaper", s.rateLimitedCheckResolverLowPriorityShaper),
+		)
+
+		rateLimitedCheckResolver := graph.NewRateLimitedCheckResolver(rateLimitConfig)
+		rateLimitedCheckResolver.SetDelegate(cycleDetectionCheckResolver)
+		s.rateLimitedCheckResolver = rateLimitedCheckResolver
```

Comment:
Outer most should still remain cycle detection. If we've detected a cycle, then there is no further work to do because we can return the error promptly. We shouldn't incur a potential dispatch delay just to find a cycle.